### ISSUE TITLE
Handling of non-standard syntax enabled by other PostCSS plugins

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#266]
cf. https://github.com/stylelint/stylelint/issues/133#issuecomment-120550968

@benfrain raised the issue that users relying on PostCSS plugins that enable non-standard syntax -- especially `postcss-nested` -- are going to have problems.

@ai, @ben-eb, and @stylelint/owners: I'm wondering if we can think of some good way to handle this problem, which I'm sure will pop-up again and again, maybe in different ways (other than nesting).

**Comment by ben-eb** [comment]
It depends on whether you think stylelint's responsibility lies with linting only *CSS* syntax, or whether it should also lint language extensions, like nested rules or mixins.

**Comment by davidtheclark** [comment]
We could also consider trying to support whatever deviant syntaxes *we can*, deciding as we go.

I just submitted a PR that might solve the nesting problem for that particular rule.

My hesitation about doing this is just that we may find we've opened a can of worms and the worms won't stop coming ... but we won't know that unless we try, right?

**Comment by MoOx** [comment]
I think we can kindly support when it's not a big deal in our codebase. Those language extensions are using css like syntax anyway, so we should be able (most of the time) to handle that easily.

**Comment by jeddy3** [comment]
> My hesitation about doing this is just that we may find we've opened a can of worms and the worms won't stop coming

This is my concern too. We've stumbled into can-of-worms issues before (particularly around where we think stylelint's responsibility lie) and, in each case, we decided to do which whatever action mitigated opening that can. Mainly for our sanity, but also for the longevity of project i.e. keeping the code-base focus on (future) css syntax rather than whatever the current flavour of the month is.

I think one option might be to go back to one of the original design decisions in #1; that the linter, like eslint, should be _pluggable_. Could this work?

Like MoOx said, I think a lot of the language extensions will work out-of-the-box as they use css-like syntax. Those that don't can be, if there's demand, be handled in plugins.

In the short term though, we do need to patch a handful of the rules to support nesting as there is a spec for it (albeit a Editor’s Draft). Fortunately, I think if we patch the rules for the spec syntax, the one `postcss-nesting` supports, then it'll also work for the `post-nested` syntax (which seems to be more widely used).

Thoughts?

**Comment by MoOx** [comment]
> Those that don't can be, if there's demand, be handled in plugins.

Yes. And postcss is going to support custom parser, so we should just keep things simple & css focused. Because with this feature, postcss is going to be much more than a simple "css parser/stringifier" but a "whatever parser to css + stringifier".

**Comment by MoOx** [comment]
Also, we may recommand people to use stylelint after non real CSS plugins.

**Comment by ben-eb** [comment]
> postcss is going to be much more than a simple "css parser/stringifier" but a "whatever parser to css + stringifier".

Or indeed, a whatever parser to whatever result. :smile:

**Comment by MoOx** [comment]
> Or indeed, a whatever parser to whatever result. :smile:

I don't think postcss should go to far. Otherwise it will become stupid... stringifier should only handle "valid" CSS ast.

**Comment by davidtheclark** [comment]
All the above seems reasonable. Right now I'm thinking that nesting and declarations inside at-rules (e.g. nested media, or mixins) might be the most significant offenders, and we've already taken care of that in a couple of places. As you've said, other plugins typically use real CSS constructs (e.g. special at-rules, special property names, special functions). But we can't predict what plugin authors might do, so we should indeed just focus on standard CSS syntax. We'll probably have to assess on a case-by-case basis any complaints saying "Using syntax enabled by plugin x broke my linting."

Does that articulate what we're thinking? Does everyone agree that that can be the approach for now?

**Comment by ben-eb** [comment]
Case by case basis seems fine to me. Indeed, I also agree that stylelint should be placed after 'pre-processing' steps but before 'post-processing' steps. So something like `precss` -> `stylelint` -> `autoprefixer` -> `cssnano` as a pipeline.

**Comment by davidtheclark** [comment]
@ben-eb If we put if after something like `precss`, won't the reported positions differ from what the user sees in her editor?

**Comment by ben-eb** [comment]
Shouldn't the positions relate back to the original source file?

**Comment by MoOx** [comment]
Positions should use `.source`. But that might be tricky instead of complicated transformation (like mixins)...

**Comment by jeddy3** [comment]
> Does that articulate what we're thinking? Does everyone agree that that can be the approach for now?

Yes, it does. I think the next thing we need to do is document this. And what I think we say depends on the following...

_BTW, this is related to the chats going on in #133  and #267_

> If we put if after something like precss, won't the reported positions differ from what the user sees in her editor?

I thought this as well, but, after I did some testing, it turns out @ben-eb is right. Odd that we've both seemingly misunderstood how the list of processors works. 

You can see this at work using `postcss-nested-props`:

Given:

```js
postcss([
  require("postcss-nested-props")(),
  require("stylelint")({
    "rules": {
      "number-leading-zero": [2, "always"],
    }
  }),
  require("postcss-reporter")()
])

```

And:

```css
a {
  font: {
    family: arial;
    size: .1em;
  };
}
```

The linter will correctly flag the warning against line 4 of the source (even though the `.1em` is on line 3 of the processed css).

So, both these pipelines are viable:

* `precss` -> `stylelint` -> `autoprefixer` -> `cssnano`
* `stylelint` -> `precss` -> `autoprefixer` -> `cssnano`

However, there seems to be advantages and disadvantages to each, and I think this is what we need to document. For example, with the former pipeline you're relying on the plugins to respect your conventions (especially your whitespace ones) e.g. `postcss-nesting` strips out the trailing semicolon, keeps the indent and whacks selectors on one line:

```css
a {
  {
    &:hover,
    &:active {
      color: pink;
    }
  }
}
```

Becomes:

```
  a:hover, a:active {
    color: pink
  }
```

Which means, given the following config:

```
"selector-list-comma-newline-after": [2, "always],
"indentation": [2, 2],
"rule-trailing-semicolon": [2, "always"]
```

You'll get no warning with the `stylelint` -> `precss` pipeline, but three `indentation`, one `selector-list-comma-newline-after` and one `rule-trailing-semicolon` warnings if you use the `precss` -> `stylelint` pipeline.

So, perhaps the user guide should read:

---

The linter supports (future) css syntax. Some _css-like_ languages extensions are, as such, supported out of the box. 

You can run the linter before or after your preprocessor steps. However, each approach has caveats:

1. Before: some plugins might require a syntax that isn't compatible with the linter. 
2. After: some plugins might generate CSS that is invalid against your config, and thus cause a warning to be flagged.

In both cases you can either turn off the offending rule, or don't use the plugin.

If you are using the linter after your preprocessor steps, you could approach the plugin author and request formatting options be added to their plugin.

---

Thoughts?

**Comment by davidtheclark** [comment]
I threw in a few modifications that I thought might help clarify the message:

The linter supports current and future CSS syntax. This includes all standard CSS but also special features that use standard CSS syntactic structures, e.g. special at-rules or special properties. Some CSS-*like* language extensions -- features that use non-standard syntactic structures -- are, as such, supported; however, since there are infinite processing possibilities, the linter cannot support everything.

You can run the linter before or after your css processors. Depending on which processors you use, each approach has caveats:

1. *Before*: Some plugins might enable a syntax that isn't compatible with the linter.
2. *After*: Some plugins might generate CSS that is invalid against your linter config, causing warnings that do not correspond to your original stylesheets.

*In both cases you can either turn off the incompatible linter rule, or stop using the incompatible plugin.* You could also approach plugin authors and request alternate formatting options that will make their plugin compatible with stylelint.

**Comment by ben-eb** [comment]
Sounds good to me. :+1:

**Comment by jeddy3** [comment]
> I threw in a few modifications that I thought might help clarify the message

I was hoping you'd do that :)

Created PR #273

**Comment by davidtheclark** [comment]
I think with that note in the README we can close this general issue, addressing specifics as they arise.
